agent/consul/state/catalog_test.go

```diff
@@ -3600,6 +3600,12 @@ func TestStateStore_CheckConnectServiceNodes_Gateways(t *testing.T) {
 	}, nil))
 	assert.True(watchFired(ws))
 
+	ws = memdb.NewWatchSet()
+	idx, nodes, err = s.CheckConnectServiceNodes(ws, "db", nil)
+	assert.Nil(err)
+	assert.Equal(idx, uint64(18))
+	assert.Len(nodes, 0)
+
 	// Watch should fire when a gateway is added
 	assert.Nil(s.EnsureService(19, "bar", &structs.NodeService{Kind: structs.ServiceKindTerminatingGateway, ID: "gateway", Service: "gateway", Port: 443}))
 	assert.True(watchFired(ws))
```

Comment:
Makes me curious why this assertion was already passing? Not sure why?

Reply:
Yeah I'm not sure what's going on there. Going to take another look at that.

It's weird that those assertions passed with the old WatchSet but failed after I added the new one.

Reply:
I wonder, if multiple watch channels trigger, does `watchFired` only find one channel that triggered. That is, is it possible this assertion found a watch that fired before `EnsureService` above?

Reply:
I figured out what happened. The channel for GatewayServices gets closed after we do the config entry update. I believe memdb closes the channel that was associated with the node we were watching.

Any calls to `watchFired(ws)` when there is a closed channel in the WatchSet will return true, since you'll always be able to read the zero value of the channel's type: https://golang.org/ref/spec#Close.

Takeaway: Any time we update a value that impacts a particular channel we should provide a new WatchSet.